packages/hooks/src/useCollectionQuery/useCollectionQuery.ts

```diff
@@ -0,0 +1,343 @@
+import {
+  ApolloError,
+  DocumentNode,
+  QueryHookOptions,
+  SubscribeToMoreOptions,
+  useQuery,
+} from "@apollo/client";
+import { cloneDeep } from "lodash";
+import { useCallback, useEffect, useState } from "react";
+import { Node, uniqueNodes } from "./uniqueNodes";
+import { Edge, createEdge, uniqueEdges } from "./uniqueEdges";
+import { useIsMounted } from "../useIsMounted";
+import { Configuration } from "../configure";
+
+interface UseCollectionQueryArguments<TQuery, TSubscription> {
+  /**
+   * The graphQL query that fetches the collection
+   */
+  query: DocumentNode;
+
+  /**
+   * A list of options for us to pass into the apollo `useQuery` hook
+   */
+  queryOptions?: QueryHookOptions<TQuery>;
+
+  /**
+   * A function that returns the location where the {@link Collection} is located.
+   *
+   * The collection is the part of the result that needs to be paginated.
+   */
+  getCollectionByPath: GetCollectionByPathFunction<TQuery>;
+
+  /**
+   *  A list of subscription options if you want to create a GraphQL
+   *  subscription to listen for more content.
+   */
+  subscription?: ListSubscription<TSubscription>;
+}
+
+interface ListSubscription<TSubscription> {
+  /**
+   * The graphQL subscription that listens for more data. This query should
+   * return a single Node that matches the data structure in
+   * `getCollectionByPath<TQuery>(...).edges.node` and
+   * `getCollectionByPath<TQuery>(...).nodes
+   */
+  document: DocumentNode;
+
+  /**
+   * A list of variables to pass into the apollo `subscribeToMore` function.
+   */
+  options?: Pick<SubscribeToMoreOptions<TSubscription>, "variables">;
+
+  /**
+   * A function that returns the location where the `Node` is located on the
+   * `TSubscription` object.
+   *
+   * It should return a single Node that matches the data structure in
+   * `getCollectionByPath<TQuery>(...).edges.node` and
+   * `getCollectionByPath<TQuery>(...).nodes
+   */
+  getNodeByPath: GetNodeByPath<TSubscription>;
+}
+
+interface Collection {
+  edges?: Edge[];
+  nodes?: Node[];
+  pageInfo: {
+    endCursor?: string | undefined;
+    hasNextPage: boolean;
+    [otherProperties: string]: unknown;
+  };
+  totalCount: number;
+  [otherProperties: string]: unknown;
+}
+
+interface CollectionQueryResult<TQuery> {
+  data: TQuery | undefined;
+  error: ApolloError | undefined;
+  loadingRefresh: boolean;
+  loadingNextPage: boolean;
+  loadingInitialContent: boolean;
+  refresh(): void;
+  nextPage(): void;
+}
+
+type GetCollectionByPathFunction<TQuery> = (
+  data: TQuery | undefined,
+) => Collection | undefined;
+
+type GetNodeByPath<TSubscription> = (
+  data: TSubscription | undefined,
+) => Node | undefined;
+
+export function useCollectionQuery<TQuery, TSubscription = undefined>({
+  query,
+  queryOptions,
+  getCollectionByPath,
+  subscription,
+}: UseCollectionQueryArguments<
```

Comment:
Apollo `useQuery` signature is:
```tsx
function useQuery(query: DocumentNode , options?: QueryHookOptions<TData, TVariables>)
```

Just to be consistent, should we move `query` out of the object?
```suggestion
export function useCollectionQuery<TQuery, TSubscription = undefined>(query: DocumentNode, {
  queryOptions,
  getCollectionByPath,
  subscription,
}: UseCollectionQueryArguments<
```

Reply:
So now that I think about it a little more, does it make sense to make the first argument just the query? Unlike `useQuery` we're doing a *alot* more than just `query`ing. We're querying, fetching more, subscribing to more, and in the future `search`ing. As a result would it make more sense to make sure each argument is labeled when it's used because the set of features that this will do is quite a bit larger than `useQuery`?

Reply:
@hosseintalebi Thoughts?